Temporary: Remove logic from DTO

Mejtika opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

## Where should the reviewer start?
<!--- Describe where reviewer should start testing -->

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How has this been tested?
<!--- Please describe in detail how you tested your changes. -->

## Screenshots (if appropriate)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Extensions/StartupExtensions.cs:
```diff
@@ -17,12 +18,45 @@
+            services.Configure<AdventSettings>(a =>
+                {
+                    string defaultDateTimeFormat = "dd-MM-yyyy";
+                    var isAdventEndDate = configuration.GetSection("Competition:EndDate").Value;
```
I suggest using _adventStart/EndDate_, because _IsAdventStart/EndDate_ suggests that value is boolean

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Extensions/StartupExtensions.cs:
```diff
@@ -17,12 +18,45 @@
+                        CultureInfo.InvariantCulture,
+                        DateTimeStyles.None, out var endDate);
+
+                    if (!(isValidStartDateTime && isValidEndDateTime))
```
let's simplify it:
```
if(!isValidStartDateTime || !isValidEndDateTime)
```

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Extensions/StartupExtensions.cs:
```diff
@@ -17,12 +18,45 @@
+
+                    if (!(isValidStartDateTime && isValidEndDateTime))
+                    {
+                        throw new InvalidOperationException(ExceptionsMessages.WrongFormatOfDate);
```
nice to add to exception parameter with default date format

Mejtika replied:
Do you suggest changing WrongFormatOfDate and add info about default date format?

WTobor replied:
Si ;) add parameter wchis is defaultFormatDate variable

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Extensions/StartupExtensions.cs:
```diff
@@ -17,12 +18,45 @@
+                throw new ArgumentNullException(nameof(configuration));
+            }
+
+            services.Configure<AdventSettings>(a =>
```
rename _a_ with _settings_

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Views/Home/Index.cshtml:
```diff
@@ -2,14 +2,17 @@
-
+@using DevAdventCalendarCompetition.Services.Options
+@using Microsoft.Extensions.Options
+@inject AdventSettings AdventSettings
```
nice <3

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Views/Home/Index.cshtml:
```diff
@@ -153,30 +156,30 @@
-        </section>
+            </section>
+        }
     }
```
what about else clause?

WTobor replied:
please find and use advent count down div

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Views/Test/Index.cshtml:
```diff
@@ -32,7 +37,7 @@
                     else
                     {
-                        if (Model.SponsorLogoUrl != null && Model.IsAdvent)
+                        if (Model.SponsorLogoUrl != null && isAdvent)
```
it's ok, but we can change the order - first check isAdvent and then check model

WTobor replied:
as I mentioned review before - let's start with `isAdvent`

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Views/Test/Index.cshtml:
```diff
@@ -41,7 +46,7 @@
                         }
 
-                        if (Model.Discount != null && Model.IsAdvent)
+                        if (Model.Discount != null && isAdvent)
```
it's ok, but we can change the order - first check isAdvent and then check model

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/appsettings.Production.json:
```diff
@@ -24,5 +24,9 @@
   },
   "DataProtection": {
     "Keys": "/etc/keys"
+  },
```
please update README

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/HomeService.cs:
```diff
@@ -1,11 +1,13 @@
 using DevAdventCalendarCompetition.Repository.Models;
+using DevAdventCalendarCompetition.Services.Extensions;
 using DevAdventCalendarCompetition.Services.Interfaces;
 using DevAdventCalendarCompetition.Services.Models;
```
why are we need that change?

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/HomeService.cs:
```diff
@@ -1,11 +1,13 @@
 using DevAdventCalendarCompetition.Services.Models;
+using DevAdventCalendarCompetition.Services.Resources;
+using Microsoft.Extensions.Configuration;
 
```
why are we need that change?

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/Models/TestDto.cs:
```diff
@@ -1,15 +1,12 @@
 using System;
 using System.Collections.Generic;
+using System.Globalization;
```
why are we need that change?

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/Profiles/TestProfile.cs:
```diff
@@ -1,5 +1,6 @@
 using AutoMapper;
 using DevAdventCalendarCompetition.Repository.Models;
+using DevAdventCalendarCompetition.Services.Extensions;
```
why are we need that change?

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/TestService.cs:
```diff
@@ -4,8 +4,12 @@
 using AutoMapper;
 using DevAdventCalendarCompetition.Repository.Interfaces;
 using DevAdventCalendarCompetition.Repository.Models;
+using DevAdventCalendarCompetition.Services.Extensions;
```
why are we need that change?

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Controllers/AccountController.cs:
```diff
@@ -1,12 +1,11 @@
 using DevAdventCalendarCompetition.Extensions;
 using DevAdventCalendarCompetition.Models.Account;
+using DevAdventCalendarCompetition.Resources;
 using DevAdventCalendarCompetition.Services.Interfaces;
```
why are we need that change?

Mejtika replied:
We're actually using it.

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Controllers/HomeController.cs:
```diff
@@ -3,10 +3,12 @@
 using DevAdventCalendarCompetition.Models.Home;
-using DevAdventCalendarCompetition.Models.Test;
 using DevAdventCalendarCompetition.Providers;
+using DevAdventCalendarCompetition.Services.Extensions;
```
why are we need that change?

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Controllers/TestController.cs:
```diff
@@ -4,6 +4,7 @@
 using System.Security.Claims;
 using DevAdventCalendarCompetition.Models.Test;
 using DevAdventCalendarCompetition.Repository.Models;
+using DevAdventCalendarCompetition.Services.Extensions;
```
why are we need that change?

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/DevAdventCalendarCompetition.csproj:
```diff
@@ -88,12 +88,11 @@
     </EmbeddedResource>
     <EmbeddedResource Update="Resources\LoggingMessages.resx">
-      <Generator>PublicResXFileCodeGenerator</Generator>
+        <Generator>PublicResXFileCodeGenerator</Generator>
```
a blank space

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Extensions/StartupExtensions.cs:
```diff
@@ -17,12 +18,45 @@
+
+                    if (!isValidStartDateTime || !isValidEndDateTime)
+                    {
+                        throw new InvalidOperationException(ExceptionsMessages.WrongFormatOfDate);
```
you are not passing defaultDateTimeFormat parameter to this resource

Mejtika replied:
![image](https://user-images.githubusercontent.com/28907927/91844188-490b2b00-ec57-11ea-9f8a-866691cbdbc7.png)

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Views/Test/Index.cshtml:
```diff
@@ -41,7 +46,7 @@
                         }
 
-                        if (Model.Discount != null && Model.IsAdvent)
+                        if (Model.Discount != null && isAdvent)
```
as I mentioned review before - let's start with `isAdvent`

Mejtika replied:
You are commenting on outdated files

WTobor reviewed (changes requested):
Please review the PR before assigning it to reviewer :) there are many changes that are not related with your work (added usings)

Mejtika commented:
I've cleaned up the solution, removed unused directives and blank spaces. I've also updated the base branch to bugfix/Remove_logic_from_TestDto. It's ready for review @WTobor.

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/Helpers/DateTimeProviderContext.cs:
```diff
@@ -0,0 +1,43 @@
+
+namespace DevAdventCalendarCompetition.Services
+{
+    public class DateTimeProviderContext : IDisposable
```
if you are using someone else's solution, it is good to practise to add a link to it (code or blog post) - so you should probably add https://dvoituron.com/2020/01/22/UnitTest-DateTime/

Mejtika replied:
Do you mean in comment in code?

WTobor replied:
yes, sth like
`//based on solution https://dvoituron.com/2020/01/22/UnitTest-DateTime/`

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/Helpers/DateTimeProviderContext.cs:
```diff
@@ -0,0 +1,43 @@
+        public void Dispose()
+        {
+            this.Dispose(true);
+            GC.SuppressFinalize(this);
```
add comment
// to prevent the dispose method from running twice

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/Helpers/DateTimeProviderContext.cs:
```diff
@@ -0,0 +1,43 @@
+            {
+                if (ThreadScopeStack.Value.Count == 0)
+                {
+                    return null;
```
See the warning - Added line #L23 was not covered by tests

Mejtika replied:
Actually, I don't see it, but I see some disadvantages to this solution. It is called the "Ambient context anti-pattern" for a reason. I'm may propose another solution. Different class for checking IsAdvent logic that has injected AdventSettings and IDateTime.

WTobor commented on CONTRIBUTING.md:
```diff
@@ -80,6 +80,10 @@ Login: `devadventcalendar@gmail.com`
 
+## Advent settings
+
+In order to imitate advent time, settings need to be changed. You can find "StartDate" and "EndDate" settings in appsettings.json under "Competition" section. We use specific DateTime format: "dd-MM-yyyy".
```
in `appsettings.Development.json`

please use markdown formatting char `

WTobor commented:
Finally! :)
